nuxeo-features/nuxeo-platform-comment/nuxeo-platform-comment/src/main/java/org/nuxeo/ecm/platform/comment/impl/CommentsMigrator.java

```diff
@@ -59,21 +73,67 @@
 
     protected static final int BATCH_SIZE = 50;
 
+    /**
+     * /**
+     *
+     * @since 11.1.
+     */
+    public final static String GET_HIDDEN_FOLDERS_QUERY = String.format("SELECT %s FROM Document WHERE %s='%s'",
+            ECM_UUID, ECM_PRIMARYTYPE, HIDDEN_FOLDER_TYPE);
+
     @Override
     protected String probeSession(CoreSession session) {
-        CommentService commentComponent =  Framework.getService(CommentService.class);
+        CommentService commentComponent = Framework.getService(CommentService.class);
         CommentServiceConfig commentServiceConfig = commentComponent.getConfig();
         if (commentServiceConfig != null) {
             Graph graph = Framework.getService(RelationManager.class).getGraph(commentServiceConfig.graphName, session);
-            if (graph.getStatements().size() > 0) {
+            if (!graph.getStatements().isEmpty()) {
                 return MIGRATION_STATE_RELATION;
             }
         }
-        return MIGRATION_STATE_PROPERTY;
+        // If not in relation, check if there are still property comments under hidden Comments folder(s)
+        // Do a new query one to check the step and the second the make the migration (do all comments)
+        if (hasUnsecuredComments(session)) {
+            return MIGRATION_STATE_PROPERTY;
+        }
+        // Else, comments are already secured
+        return MIGRATION_STATE_SECURED;
+    }
+
+    /**
+     * @since 11.1
+     */
+    protected List<DocumentModel> getUnsecuredComments(CoreSession session) {
+        return getCommentFolders(session).stream()
+                                         .flatMap(f -> session.getChildren(f.getRef(), COMMENT_DOC_TYPE).stream())
+                                         .collect(Collectors.toList());
+    }
+
+    /**
+     * @since 11.1
+     */
+    protected boolean hasUnsecuredComments(CoreSession session) {
+        List<DocumentModel> folders = getCommentFolders(session);
+        return folders.stream().anyMatch((folder) -> session.hasChildren(folder.getRef()));
+    }
+
+    /**
+     * @since 11.1
+     */
+    protected List<DocumentModel> getCommentFolders(CoreSession session) {
+        return session.query(GET_HIDDEN_FOLDERS_QUERY, (doc) -> doc.getName().equals(COMMENTS_DIRECTORY)
```

Comment:
Here you're using post filtering API but this is not performant. On a very large repository, there could be several `HIDDEN_FOLDER_TYPE `. We need to find a more performant query to get desired comments.

Reply:
yes we agree, it's why we make a new commit which proposal is to use the scrolling, i don't know wdty, plus that the T&P on mongo failed, https://qa.nuxeo.org/jenkins/job/TestAndPush/job/ondemand-testandpush-nalkotob/44/ .

another way of we can't using the scroll solution is to get `Comments` folder in a separate query and then get all its child as below :

Reply:
It should work, note that you can use a NXQL to get the children once you have all parent ids.

This will have better performance instead of fetching them one by one, furthermore with queryProjection you can just get what you need which it should be the id in order to move it.

Reply:
when you said it should work, you are talking about the solution with the use of `while` loop... ?

Reply:
when you said it should work, you are talking about the solution with the use of `while` loop... ? 

About the use of NXQL to get the children of given parent Ids: are you talking about using the `IN` operator in the NXQL ? If it's the case, we tried it at the start of the migration development but we didn't find a way to do the `SubQuery IN (select...)` and there is no way to do a subquery in NXQL as far as we know. On the other hand, we know that we can query in (parent1, parent2...) but we didn't find a pretty way to do it so far unless we use the StringBuilder to build the id list (make concatenations ....)

Reply:
The logic may work, the loop is to avoid due to performance.

Indeed there's no sub query in NXQL.

Once you have the parent folder id, you can query its direct children with:
```
SELECT * FROM Document WHERE ecm:parentId IN (...)
```
Or all children with:
```
SELECT * FROM Document WHERE ecm:ancestorId IN (...)
```

To build the list of ancestorIds you can use streams:
```java
parentIds.stream().map(id -> "'" + id + "'").collect(Collectors.joining(", "));
```
which will produces: _'id1', 'id2', ..., 'idn'_

Or:
```java
parentIds.steam().collect(Collectors.joining(", ", "('", "')");
```
which will produces: _('id1', 'id2', ..., 'idn')_

Reply:
Check that the list of `parentIds` is not empty before doing that though.

Reply:
We end up by using a NXQL with the `IN` operator : 
- To get all comments folders `Ids` (Comments under Domain + the Sub folders under these Comments) using session#queryProjection
- To get all comments documents, unfortunately we need the comment  document model to move it from the current path to the new one, 
` session.move(comment.getRef(), commentManager.getAncestorRef(session, comment), null);`  i don't know if we can get the AncestorIds (`comment:ancestorIds`) in projection our idea is to replace the call on `commentManager.getAncestorRef` and if not what is the best way to get this document model directly by using `session#query` or by getting the document unitary `session#getDocument` (then it will be use the session cache if its loaded before, but we can end up with memory issu if there is many comments) 

another question, we want to use a projection `count(ecm:uuid)`to check if there isa unsecured comment, but we think it will not be work on  [`mongoDB`](https://qa.nuxeo.org/jenkins/job/TestAndPush/job/ondemand-testandpush-nalkotob/44/ )

Reply:
We endup by creating a projection query that give us the `ecm:uuid` of the comment document and his top level `ancestor` (the file being commented) but unfortunately. We didn't put this code, because in our migration we need the document model for two propos :

Reply:
We endup by creating a projection query that give us the ecm:uuid of the comment document and his top level ancestor (the file being commented) but unfortunately. We didn't put this code, because in our migration we need the document model for two propos :

Reply:
We made the query with projection that give (`ecm:uuid` and `ancestor`), but our migration need the document model to : StripACL and disabling notification event.